### DISCUSSION TITLE
Anyone interested in an icons library?

### DISCUSSION THREAD

**Discussion by thednp** [vanjs-org/van#326] [Show and tell]
I'm working on a library for VanJS...

![Screenshot from 2024-05-19 22-14-14](https://github.com/vanjs-org/van/assets/6761246/f1d7e72c-e6d1-4b38-9acf-c3304c14c03e)

It's basically in tooling setup, coding complete...

Anyone interested :smiley: ?

**Comment by thednp** [comment]
@Tao-VanJS 

[Demo](https://thednp.github.io/vanjs-feather)  -  [Repo](https://github.com/thednp/vanjs-feather)

Btw, I couldn't find where to add this to the [community-add-ons](https://vanjs.org/#community-add-ons) section, can you please help with that?

**Reply by sirenkovladd** [reply]
https://github.com/vanjs-org/vanjs-org.github.io/blob/master/sitegen/home.ts#L131-L187

**Comment by thednp** [comment]
Hi @Tao-VanJS, @sirenkovladd and everybody else!

I'm now thinking about a 1.0 release, but I want to make sure some things are ironed out.
* First and most importantly I want to have this library work with SSR if possible, so any suggestion on how to implement is greatly appreciated. I would really want SSR powered demo for VanJSFeather at some point.
* Secondly, I want to make sure it works with `vanjs-jsx`, I know a lot of people are going to ask and for good reason (weird paradox JSX makes for better DX for some people, but with VanJS you actually write much less code, and for some reason TailwindCSS Intellisense won't work with the "class" property of JS objects).

Also I need to know your opinions as well, please have a look at the code, give it a try,maybe there would be some room for improvement, really any feedback is welcome, greatly appreciated! Thank you!

**Reply by Tao-VanJS** [reply]
> * First and most importantly I want to have this library work with SSR if possible, so any suggestion on how to implement is greatly appreciated. I would really want SSR powered demo for VanJSFeather at some point.

You can refer to section "Shared UI Components" of page "SSR & Hydration" in the VanJS website on how to build components that can be used by both SSR and CSR.

> * Secondly, I want to make sure it works with `vanjs-jsx`, I know a lot of people are going to ask and for good reason (weird paradox JSX makes for better DX for some people, but with VanJS you actually write much less code, and for some reason TailwindCSS Intellisense won't work with the "class" property of JS objects).

@cqh963852 is the owner of `vanjs-jsx`, so I will let him comment.

**Reply by cqh963852** [reply]
I think vanjs should be supported first.

van_jsx just demonstrates the jsx capabilities with vanjs as the core.

---

There is still a lot of uncertainty and van_jsx is not production ready at the moment.

For example

- the optimization of reconcile.
- How to write jsx attributes, use `class` or `className`?

**Reply by cqh963852** [reply]
Consider van_jsx compatibility. 

We can use code generation to generate van_jsx compatible code based on the code your contribute.

**Reply by thednp** [reply]
> You can refer to section "Shared UI Components" of page "SSR & Hydration" in the VanJS website on how to build components that can be used by both SSR and CSR.

@Tao-VanJS 
I'm thinking I need to create another file for SSR where the [FeatherIcon](https://github.com/thednp/vanjs-feather/blob/438e2593cfbc883e73306db83fb351190ea70b2f/src/index.ts#L16C7-L16C18) template uses the van object from a parent element, but I have no idea how it could work dynamically depending on the render type (client / server). Any idea?

**Reply by Tao-VanJS** [reply]
> I'm thinking I need to create another file for SSR where the [FeatherIcon](https://github.com/thednp/vanjs-feather/blob/438e2593cfbc883e73306db83fb351190ea70b2f/src/index.ts#L16C7-L16C18) template uses the van object from a parent element, but I have no idea how it could work dynamically depending on the render type (client / server). Any idea?

I don't quite understand the question.

**Reply by thednp** [reply]
Can you please make an example component based on FeatherIcon to be used in a Home component that would work on both client and  server?

**Reply by Tao-VanJS** [reply]
Could it be something like that:

```ts
type IconProps = {
  van: VanObj
  <Your other props>
}

const FeatherIcon(props: IconProps, ...children: Element[]) => {
  ...
}

type MyComponentProps {
  van: VanObj
  <Other props>
}

const MyComponent = (props: MyComponentProps) => {
  return FeatherIcon({van: props.van, ...}, ...)
}
```

**Reply by sirenkovladd** [reply]
@thednp I think this change can help you make
> dynamically depending on the render type (client / server)

https://github.com/vanjs-org/mini-van/pull/10

**Reply by thednp** [reply]
That would be an awesome addition if indeed works :rescue_worker_helmet:

**Reply by Tao-VanJS** [reply]
Hi @thednp,

[Mini-Van 0.6.0](https://github.com/vanjs-org/mini-van/releases/tag/0.6.0) was just released, which improves the developer experience of fullstack rendering. Among other things, it allows `van` object to be registered via `registerEnv` so that the `van` object doesn't need to be passed around shared components via parameters.

**Reply by thednp** [reply]
Thank you @Tao-VanJS will have a look over the weekend. :ok: 

Now the next thing is to create a vite starter project with full stack and SSR support. Unfortunately due to some data loss my starter I was working on is gone, but I hope I can redo all that quickly. Something for [create vite extra](https://github.com/bluwy/create-vite-extra) I'm thinking would be best.
What do you think?

**Reply by thednp** [reply]
@Tao-VanJS please tell me what do you think about [this implementation](https://github.com/thednp/vanjs-feather/blob/main/src/FeatherIcon.ts)?

**Reply by Tao-VanJS** [reply]
Yeah, that implementation looks good to me :-)

**Reply by thednp** [reply]
What if we create a simple vite plugin that does this automatically?

```ts
import vanPlate from "mini-van-plate/van-plate";
import vanCore, { State } from "vanjs-core";

/* istanbul ignore next @preserve */
if (typeof window === "undefined") {
  registerEnv({ van: vanPlate });
} else {
  registerEnv({ van: vanCore });
}
```

SPA apps won't be affected, while bot SPA and SSR apps will have access to the van they like to work with.

**Reply by thednp** [reply]
> What if we create a simple vite plugin that does this automatically?
> 
> ```ts
> import vanPlate from "mini-van-plate/van-plate";
> import vanCore, { State } from "vanjs-core";
> 
> /* istanbul ignore next @preserve */
> if (typeof window === "undefined") {
>   registerEnv({ van: vanPlate });
> } else {
>   registerEnv({ van: vanCore });
> }
> ```
> 
> SPA apps won't be affected, while bot SPA and SSR apps will have access to the van they like to work with.

@Tao-VanJS plugin complete and ready for testing. Next up is a good router to work with this stuff.
